index.bs

```diff
@@ -3209,6 +3210,84 @@ opened</dfn> steps given |window|, |type| and |message|.
 
 </div>
 
+## Emulation ## {#module-emulation}
+
+The <dfn export for=modules>emulation</dfn> module contains functionality for
+emulating device-dependent and system-dependent parameters.
+
+### Definition ### {#module-emulation-definition}
+
+[=remote end definition=]
+
+<pre class="cddl remote-cddl">
+
+EmulationCommand = (
+  emulation.SetViewport
+)
+</pre>
+
+### Types ### {#module-emulation-types}
+
+### Commands ### {#module-emulation-commands}
+
+#### The emulation.setViewport Command ####  {#command-emulation-setViewport}
+
+The <dfn export for=commands>emulation.setViewport</dfn> command turns the emulation of the viewport on.
+
+<dl>
+   <dt>Command Type</dt>
+   <dd>
+    <pre class="cddl remote-cddl">
+      emulation.SetViewport = {
+        method: "emulation.setViewport",
+        params: emulation.SetViewportParameters
+      };
+
+      emulation.SetViewportParameters = {
+        context: browsingContext.BrowsingContext,
+        ? viewport: emulation.Viewport
+      }
+
+      emulation.Viewport = {
+        width: js-uint,
+        height: js-uint,
+      }
+    </pre>
+   </dd>
+   <dt>Return Type</dt>
+   <dd>
+    <pre class="cddl">
+     EmptyResult
+    </pre>
+   </dd>
+</dl>
+
+<div algorithm="remote end steps for emulation.setViewport">
+
+The [=remote end steps=] with |command parameters| are:
+
+1. Let |context id| be the value of the <code>context</code> field of
+   |command parameters|.
+
+1. Let |context| be the result of [=trying=] to [=get a browsing context=] with |context id|.
+
+1. If |context| is not a [=top-level browsing context=], return [=error=] with [=error code=] [=unsupported operation=].
+
+1. If |command parameters| [=map/contains=] "<code>viewport</code>", then:
+    1. Let |viewport parameters| be the <code>viewport</code> attribute of |command parameters|.
+    1. If |viewport parameters| [=map/contains=] "<code>width</code>" and |viewport parameters| [=map/contains=] "<code>height</code>", then:
+        1. Let |width| be the <code>width</code> attribute of |viewport parameters| in CSS pixels.
+        1. Let |height| be the <code>height</code> attribute of |viewport parameters| in CSS pixels.
+        1. If the |width| or |height| is outside of the implementation-specific bounds, return an [=error=] with the [=error code=] [=invalid argument=].
+        1. Change the layout viewport of the |context| to be |width| pixels wide and |height| pixels high and run the [[cssom-view-1#resizing-viewports]] steps.
+    1. Else:
+        1. Return [=error=] with [=error code=] [=invalid argument=].
+1. Else:
+    1. Reset any previously emulated viewport on the |context|.
```

Comment:
How does this work with several sessions? Let's say session 1 applied a viewport emulation and session 2 then applied another one. If session 2 "resets" the viewport emulation, I imagine we reset to "no emulation at all", and not just to session 1's emulation?

(and in general I wonder if the "reset" needs to be more explicit, eg which values to we reset to, but I'll let folks more used to writing spec comment on that)

Reply:
Good point. Since the viewport is essentially a shared global state, I think we should just allow the most recent command to "win", so if a session resets the viewport, it is reset even if some other session has set it previously. I am not sure how to express it in the spec: I have rephrased a bit, PTAL.

Reply:
Yeah, I think the viewport size is inevitably shared global state. I can't really think of a good way to avoid conflicts. Maybe the command could fail if there's already an override set by a different session? Although that's a bit tricky as it involves cross-session state and can leave you in a state that you can't get out of (if the other session closes).

Reply:
There is another edge case around other clients (that BiDi has not way of knowing about) emulating the viewport: e.g., opening DevTools might change the viewport size. I think that having the command apply the specified viewport even if there are other sessions trying to do the same is the least surprising behaviour.